File: code/clojure/rest/server/src/task_server/task.clj
(ns task-server.task
  (:require [clj-time.core :as time]
            [clj-time.format :as time-format]))

(def ^:private iso-formatter (time-format/formatters :date-time))

(defn- current-time-str []
  (time-format/unparse iso-formatter (time/now)))

(defn- generate-uuid []
  (str (java.util.UUID/randomUUID)))

;; Task status constants
(def task-statuses #{:pending :in-progress :completed :cancelled})

;; Task priority constants  
(def task-priorities #{:low :medium :high :urgent})

(defn create-task
  "Create a new task with default values"
  [{:keys [title description status priority tags assigned-to]
    :or {description ""
         status :pending
         priority :medium
         tags []
         assigned-to ""}}]
  (let [now (current-time-str)]
    {:id (generate-uuid)
     :title title
     :description description
     :status status
     :priority priority
     :tags tags
     :assigned-to assigned-to
     :created-at now
     :updated-at now}))

(defn update-task
  "Update an existing task with new values"
  [task updates]
  (let [updated-task (merge task 
                           (select-keys updates [:title :description :status 
                                                :priority :tags :assigned-to]))]
    (assoc updated-task :updated-at (current-time-str))))

(defn valid-status? [status]
  (contains? task-statuses (keyword status)))

(defn valid-priority? [priority]
  (contains? task-priorities (keyword priority)))

(defn validate-task
  "Validate task data"
  [{:keys [title status priority]}]
  (cond
    (or (nil? title) (empty? title))
    {:valid? false :error "Title is required"}
    
    (and status (not (valid-status? status)))
    {:valid? false :error "Invalid status"}
    
    (and priority (not (valid-priority? priority)))
    {:valid? false :error "Invalid priority"}
    
    :else
    {:valid? true}))

File: code/elixir/rest/server/lib/task_server/task.ex
defmodule TaskServer.Task do
  @moduledoc """
  Task model representing a task in the system
  """

  @derive Jason.Encoder
  defstruct [
    :id,
    :title,
    :description,
    :status,
    :priority,
    :tags,
    :assigned_to,
    :created_at,
    :updated_at
  ]

  @type status :: :pending | :in_progress | :completed | :cancelled
  @type priority :: :low | :medium | :high | :urgent

  @type t :: %__MODULE__{
    id: String.t(),
    title: String.t(),
    description: String.t(),
    status: status(),
    priority: priority(),
    tags: [String.t()],
    assigned_to: String.t(),
    created_at: DateTime.t(),
    updated_at: DateTime.t()
  }

  @doc """
  Creates a new task with default values
  """
  def new(attrs \\ %{}) do
    now = DateTime.utc_now()
    
    %__MODULE__{
      id: UUID.uuid4(),
      title: Map.get(attrs, "title", Map.get(attrs, :title, "")),
      description: Map.get(attrs, "description", Map.get(attrs, :description, "")),
      status: parse_status(Map.get(attrs, "status", Map.get(attrs, :status, "pending"))),
      priority: parse_priority(Map.get(attrs, "priority", Map.get(attrs, :priority, "medium"))),
      tags: Map.get(attrs, "tags", Map.get(attrs, :tags, [])),
      assigned_to: Map.get(attrs, "assigned_to", Map.get(attrs, :assigned_to, "")),
      created_at: now,
      updated_at: now
    }
  end

  @doc """
  Updates a task with new attributes
  """
  def update(task = %__MODULE__{}, attrs) do
    task
    |> Map.merge(attrs)
    |> Map.put(:updated_at, DateTime.utc_now())
  end

  @doc """
  Converts task to JSON-encodable map
  """
  def to_map(task = %__MODULE__{}) do
    %{
      id: task.id,
      title: task.title,
      description: task.description,
      status: to_string(task.status),
      priority: to_string(task.priority),
      tags: task.tags,
      assigned_to: task.assigned_to,
      created_at: DateTime.to_iso8601(task.created_at),
      updated_at: DateTime.to_iso8601(task.updated_at)
    }
  end

  defp parse_status(status) when is_atom(status), do: status
  defp parse_status("pending"), do: :pending
  defp parse_status("in_progress"), do: :in_progress
  defp parse_status("completed"), do: :completed
  defp parse_status("cancelled"), do: :cancelled
  defp parse_status(_), do: :pending

  defp parse_priority(priority) when is_atom(priority), do: priority
  defp parse_priority("low"), do: :low
  defp parse_priority("medium"), do: :medium
  defp parse_priority("high"), do: :high
  defp parse_priority("urgent"), do: :urgent
  defp parse_priority(_), do: :medium
end